src/Oscoin/Storage/Block/Cache.hs

```diff
@@ -0,0 +1,65 @@
+{-- | A cache for 'Blocks', implemented as a ring buffer.
+ -- It caches the \"mutable\" part of the chain, which are usually the
+ -- last @k@ blocks, where @k@ is a protocol constant.
+ --}
+module Oscoin.Storage.Block.Cache
+    ( BlockCache
+    , newBlockCache
+    , consBlock
+    , viewBlocks
+    , invalidate
+    , cached
+    ) where
+
+import           Oscoin.Prelude hiding (length)
+
+import           GHC.Natural
+
+import           Oscoin.Crypto.Blockchain.Block (Block)
+
+import qualified Data.Foldable as F
+import           Data.Sequence.Circular (CircularSeq)
+import qualified Data.Sequence.Circular as CSeq
+
+newtype BlockCache c tx s =
+    BlockCache { getBlockCache :: MVar (CircularSeq (Block c tx s)) }
+
+cached
+    :: BlockCache c tx s
+    -> Int
+    -- ^ The number of requested elements
+    -> (Int -> IO [Block c tx s])
+    -- ^ A function to \"backfill\" the cache by the number of missing
+    -- elements.
+    -> IO [Block c tx s]
```

Comment:
How about using `Oscoin.Time.Chrono` here, too? We should point out that we request the X most recently added elements.

Reply:
I've deliberately tried to go light with the usage of `Chrono` to not make this PR even beefier. Do you think we should bite the bullet or perhaps we can spin up a follow-up ticket, so that the next PR is way leaner?

Reply:
Yes, it’s probably a good idea to leave this for later but please document the ordering